### ISSUE TITLE
Extended use leads to Nova hang

### ISSUE THREAD

**Issue by gingerbeardman** [robochase6000/navigation.novaextension#2]
- use Navigation for a long session
- eventually Nova will "hang" whilst saving a doc
- Nova remains responsive but the save never completes

disaster avoidance
- copy the current text to the clipboard
- force quit nova
- restart nova
- (nova tells me that Navigation extension had problems)
- paste copied text into document to make sure you don't lose edits

**Comment by chase-rg** [comment]
thanks for reporting, and sorry about that @gingerbeardman 

how long of a session are you talking?

can you describe how the session goes leading up to the hang?  does it run slower and slower as you go, or is everything fine and then it just hangs out of the blue?

I'll look into this, I haven't had had time for longer coding sessions since I wrote the plugin unfortunately, so I probably just haven't experienced this yet.

also, what is your history size set to in the plugin Settings?

**Comment by gingerbeardman** [comment]
Probably a day or so session, but I leave the editor open most of the time so it could have been more.

The app didn't seem to slow down, I save frequently and one save just fails to complete.

This has happened a few times so I figure it best that I reported it.

You could probably automate a session with a lot of scrolling and clicking and see what happens?

I was using the extension on default settings.

**Comment by robochase6000** [comment]
Ok thanks for the update; I swept through the code and tried to clean up a few things, but nothing stuck out to me as especially dangerous. 

I've also always been a little bit weary about how the Sidebar GUI is updated.  I took some care to try to optimize this a bit, and I've also added a couple Preferences for just turning off updating the Sidebar altogether.  I think the extension is still quite usable without the GUI personally.

Not that I can ask you to be a guinea pig here, and I understand if you don't want to keep using this extension.  But if you do, please let me know if you run into more problems.  I'm also curious on the exact messaging Nova is giving you about it.

Hopefully I can get back to coding and using the extension a bit more this week to see if I can repro your issue...

**Comment by robochase6000** [comment]
these are the new preferences btw - "Sidebar Updates Enabled", and "Auto Reveal Focused Index"

<img width="809" alt="image" src="https://github.com/robochase6000/navigation.novaextension/assets/234673/130a6fa8-1fa7-466e-bd0e-d0f54dd8d4ee">

toggling off "Sidebar Updates Enabled" should be sufficient to stop updating the Sidebar GUI.

I've wondered if there's some overhead with promises in Javascript that I'm just not used to, so turning off "Auto Reveal Focused Index" should keep things lighter, at the expense of usability

**Comment by gingerbeardman** [comment]
I've reactivated the extension, default settings, and will take a screen shot if/when this happens again.

**Comment by gingerbeardman** [comment]
Here we go.

The curious about this one is that it was a result of Nova loading and remaining unused!

1. Nova was running
1. macOS update started
1. I went to bed
1. in the morning, computer had restarted and apps reopened to the previous state
1. Nova was open with my project
1. but it was frozen (not responding according to Force Quit/Activity Monitor)
1. relaunching Nova generated a few crash logs [Nova-extension-crashes.zip](https://github.com/robochase6000/navigation.novaextension/files/12161175/Nova-extension-crashes.zip)
1. and the following "Extension Error" dialog, which is what I had seen before

Sadly, I cannot help any further with this.

![Screen shot 2023-07-25 at 13 32 57](https://github.com/robochase6000/navigation.novaextension/assets/49612/bf33f369-0aec-4a54-8b00-e3f4be699e04)

**Comment by robochase6000** [comment]
OK, I understand, thanks for the info.  Unfortunately Nova crashes for me at least once a day even before this extension existed, so it's tough for me to keep a session going for more than a few hours.  Anyways hopefully a fix will emerge eventually, thanks for your time :)

**Comment by gingerbeardman** [comment]
Yes, I feel and understand the pain.

I also get lots of other crashes. I've managed to get a couple solved by emailing Nova support but there are many remaining...

**Comment by gingerbeardman** [comment]
Also reported to Nova.

**Comment by robochase6000** [comment]
ok cool, I wonder if they'll have any leads.  have you experienced more problems?  

Since my previous message here, I have pushed a couple null reference fixes, and I've also logged considerable time coding my own project while this plugin is active.  I'd guess around 20 hours of work.  Nova crashed once in that time and it didn't seem related to this plugin, but otherwise things have been pretty stable on my end with the default settings

**Comment by gingerbeardman** [comment]
> have you experienced more problems?

Yes, multiple with extensions and also multiple other Nova bugs. More than I have time to devote to. So I have to pick my battles.

**Comment by robochase6000** [comment]
I finally got this to happen!  unfortunately there was nothing obvious I could garner. 

in my case, I had the playdate simulator running for quite a long time, and then I quit the playdate simulator.  from there, Nova seemed to think the playdate simulator was still running - I could press the Stop button and interact with the editor a bit, but it didn't seem able to refresh the visual state of the Stop button. I didn't play around with the editor much at that point; I just quit and restarted, and Nova said the Navigation plugin had a problem.  

Not super helpful unfortunately, but I guess now I know how the editor behaves when you get into this state :/

**Comment by gingerbeardman** [comment]
Any progress is good progress! Thanks for looking out for it.
